File: CarHomeWidget/src/main/java/com/anod/car/home/app/AppsListAdapter.java
package com.anod.car.home.app;

import com.anod.car.home.model.AppsList;
import com.anod.car.home.utils.UtilitiesBitmap;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.drawable.BitmapDrawable;
import android.net.Uri;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.ImageView;
import android.widget.TextView;

import java.util.ArrayList;

import butterknife.ButterKnife;
import butterknife.BindView;

public class AppsListAdapter extends ArrayAdapter<AppsList.Entry> {
    final private int mResource;
    final private Context mContext;
    private final AppIconLoader mIconLoader;
    private final BitmapDrawable mDefaultIconDrawable;

    public AppsListAdapter(Context context, int resource, AppIconLoader iconLoader) {
        super(context, resource, new ArrayList<AppsList.Entry>());
        mResource = resource;
        mContext = context;
        Bitmap defaultIcon = UtilitiesBitmap.makeDefaultIcon(context.getPackageManager());
        mDefaultIconDrawable = new BitmapDrawable(context.getResources(), defaultIcon);
        mIconLoader = iconLoader;
    }

    @Override
    public View getView(int position, View view, ViewGroup parent) {
        ViewHolder holder;
        if (view != null) {
            holder = (ViewHolder) view.getTag();
        } else {
            LayoutInflater inflater = (LayoutInflater) mContext
                    .getSystemService(Context.LAYOUT_INFLATER_SERVICE);
            view = inflater.inflate(mResource, parent, false);
            holder = new ViewHolder(view);
            view.setTag(holder);
        }

        AppsList.Entry entry = getItem(position);

        holder.title.setText(entry.title);

        if (entry.componentName == null) {
            if (entry.iconRes > 0) {
                holder.icon.setVisibility(View.VISIBLE);
                mIconLoader.picasso()
                        .load(entry.iconRes)
                        .placeholder(mDefaultIconDrawable)
                        .into(holder.icon);
            } else {
                holder.icon.setVisibility(View.INVISIBLE);
            }
        } else {
            holder.icon.setVisibility(View.VISIBLE);
            mIconLoader.picasso()
                    .load(Uri.fromParts(AppIconLoader.SCHEME,entry.componentName.flattenToShortString(),null))
                    .placeholder(mDefaultIconDrawable)
                    .into(holder.icon);
        }
        view.setId(position);
        return view;
    }


    static class ViewHolder {

        int position;

        @BindView(android.R.id.text1)
        TextView title;

        @BindView(android.R.id.icon)
        ImageView icon;

        public ViewHolder(View view) {
            ButterKnife.bind(this, view);
        }
    }
}

File: app/src/main/java/com/anod/car/home/app/AppsListAdapter.kt
package com.anod.car.home.app

import android.content.Context
import android.graphics.drawable.BitmapDrawable
import android.net.Uri
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.ImageView
import android.widget.TextView
import com.squareup.picasso.Picasso
import info.anodsplace.carwidget.chooser.ChooserEntry
import info.anodsplace.carwidget.content.graphics.PackageIconRequestHandler
import info.anodsplace.carwidget.content.graphics.UtilitiesBitmap
import java.util.*

class AppsListAdapter(
        context: Context,
        private val resource: Int,
        private val iconLoader: Picasso
) : ArrayAdapter<ChooserEntry>(context, resource, ArrayList()) {

    private val defaultIconDrawable: BitmapDrawable

    init {
        val defaultIcon = UtilitiesBitmap.makeDefaultIcon(context.packageManager)
        defaultIconDrawable = BitmapDrawable(context.resources, defaultIcon)
    }

    override fun getView(position: Int, view: View?, parent: ViewGroup): View {
        var itemView = view
        val holder: ViewHolder
        if (itemView != null) {
            holder = itemView.tag as ViewHolder
        } else {
            val inflater = context
                    .getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
            itemView = inflater.inflate(resource, parent, false)
            holder = ViewHolder(itemView)
            itemView!!.tag = holder
        }

        val entry = getItem(position)

        holder.title.text = entry!!.title

        if (entry.componentName == null) {
            if (entry.iconRes > 0) {
                holder.icon.visibility = View.VISIBLE
                iconLoader
                        .load(entry.iconRes)
                        .placeholder(defaultIconDrawable)
                        .into(holder.icon)
            } else {
                holder.icon.visibility = View.INVISIBLE
            }
        } else {
            holder.icon.visibility = View.VISIBLE
            iconLoader
                    .load(Uri.fromParts(PackageIconRequestHandler.SCHEME, entry.componentName!!.flattenToShortString(), null))
                    .placeholder(defaultIconDrawable)
                    .into(holder.icon)
        }
        itemView.id = position
        return itemView
    }


    class ViewHolder(itemView: View): androidx.recyclerview.widget.RecyclerView.ViewHolder(itemView) {
        var title: TextView = itemView.findViewById(android.R.id.text1)
        var icon: ImageView = itemView.findViewById(android.R.id.icon)
    }
}
